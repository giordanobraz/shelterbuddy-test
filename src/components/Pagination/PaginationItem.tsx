import { IconButton, IconButtonProps, styled } from "@mui/material";

const CustomButton = styled(IconButton)<IconButtonProps>({
  width: 40,
  height: 40,
  margin: 4,
  fontSize: 16,
  fontWeight: 600,
  borderRadius: "50%",
  backgroundColor: "#E0E7FD",
  color: "#002fd6",
  [`&:disabled`]: {
    backgroundColor: "#002fd6",
    color: "white",
  },
});

interface PaginationItemProps {
  number: number;
  isCurrent?: boolean;
  onPageChange: (page: number) => void;
}

export function PaginationItem({
  number,
  isCurrent,
  onPageChange,
}: PaginationItemProps) {
  if (isCurrent) {
    return <CustomButton disabled>{number + 1}</CustomButton>;
  }

  return (
    <CustomButton onClick={() => onPageChange(number)}>
      {number + 1}
    </CustomButton>
  );
}
