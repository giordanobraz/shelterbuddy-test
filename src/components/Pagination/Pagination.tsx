import { Stack, useMediaQuery } from "@mui/material";

import { PaginationItem } from "./PaginationItem";

export interface PaginationProps {
  totalRegisters: number;
  rowsPerPage: number;
  currentPage: number;
  setCurrentPage: (page: number) => void;
}

export function Pagination({
  totalRegisters,
  rowsPerPage,
  currentPage,
  setCurrentPage,
}: PaginationProps) {
  const mediaQuery = useMediaQuery("(max-width:810px)");

  const totalPageCount = Math.ceil(totalRegisters / rowsPerPage);

  const range = (start: number, end: number) => {
    let length = end - start + 1;
    return Array.from({ length }, (_, i) => i + start);
  };

  const pages = range(0, totalPageCount - 1);

  return (
    <Stack direction="row" overflow={mediaQuery ? "scroll" : "hidden"}>
      {pages.map((page) => (
        <PaginationItem
          key={page}
          number={page}
          isCurrent={page === currentPage}
          onPageChange={setCurrentPage}
        />
      ))}
    </Stack>
  );
}
