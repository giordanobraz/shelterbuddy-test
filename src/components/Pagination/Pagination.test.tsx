import { render, screen } from "@testing-library/react";

import { ComponentProps } from "react";
import { Pagination } from "./Pagination";
import userEvent from "@testing-library/user-event";

describe("Pagination", () => {
  it("should render", () => {
    const { container } = getRenderer();
    expect(container).toMatchSnapshot();
  });

  it("should return an array with 10 buttons", () => {
    getRenderer();
    expect(screen.getAllByRole("button")).toHaveLength(10);
  });

  it("should click on button and call setCurrentPage callback", () => {
    const setCurrentPage = jest.fn();
    getRenderer({ setCurrentPage });
    expect(setCurrentPage).not.toBeCalled();
    userEvent.click(screen.getByRole("button", { name: "2" }));
    expect(setCurrentPage).toBeCalledTimes(1);
  });
});

function getRenderer({
  currentPage = 0,
  setCurrentPage = jest.fn(),
  rowsPerPage = 10,
  totalRegisters = 100,
}: Partial<ComponentProps<typeof Pagination>> = {}) {
  return render(
    <Pagination
      currentPage={currentPage}
      setCurrentPage={setCurrentPage}
      rowsPerPage={rowsPerPage}
      totalRegisters={totalRegisters}
    />
  );
}
