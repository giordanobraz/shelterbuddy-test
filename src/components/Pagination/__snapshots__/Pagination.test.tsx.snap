// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should render 1`] = `
<div>
  <div
    class="css-1u96uvg-MuiStack-root"
  >
    <button
      class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-1e2s9g8-MuiButtonBase-root-MuiIconButton-root"
      disabled=""
      tabindex="-1"
      type="button"
    >
      1
    </button>
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1e2s9g8-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      2
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1e2s9g8-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      3
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1e2s9g8-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      4
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1e2s9g8-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      5
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1e2s9g8-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      6
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1e2s9g8-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      7
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1e2s9g8-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      8
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1e2s9g8-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      9
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1e2s9g8-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      10
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
